### ISSUE TITLE
Wrong network mask for floating ip address on login VMs

### ISSUE THREAD

**Issue by dchirikov** [clustervision/trinity#647]
Floating ip has /32 prefix witch is wrong. Routing is not optimal. Every packet from login node needs to go via 10.141.255.254
In `cv_install_first_cluster`
```
FIRST_FLOATING="172.16.255.254"
nova floating-ip-bulk-create --pool ${FIRST_POOL} ${FIRST_FLOATING}
```
/32 is the default mask assigned when you are not specifying it.
In OpenStack documentation:
```
nova floating-ip-bulk-create --pool test 192.168.1.56/29
```
http://docs.openstack.org/user-guide-admin/cli_admin_manage_ip_addresses.html

in the `/trinity_api/src/api.py`:
```
    payload={
      "floating_ips_bulk_create":{
        "ip_range":"172."+second_octet+".255.254",
        "pool": login_pool
      }
```
In OpenStack API examples prefix is being specified.
```
<?xml version='1.0' encoding='UTF-8'?>
<floating_ips_bulk_create>
  <interface>eth0</interface>
  <ip_range>192.168.1.0/24</ip_range>
  <pool>nova</pool>
</floating_ips_bulk_create>
```
https://github.com/ytsarev/openstack-nova/blob/master/doc/api_samples/os-floating-ips-bulk/floating-ips-bulk-create-resp.xml

**Comment by hansthen** [comment]
I am not sure the '/32' prefix is wrong. Can you give supporting argumentation? I will give you my reasons why I think it is correct.

For one, the prefix specified using floating-ip-bulk-create only indicate the range of ip addresses created. It has no effect on on the netmask specified when attaching a floating IP to the instance. I bulk created a range of /24 for the 172.17.0.0/24 and the end result is the same:
```
    inet 172.16.255.254/32 scope global eth1
       valid_lft forever preferred_lft forever
```
 
Second the /32 prefix is hard-coded in openstack for binding floating ips. So either the OpenStack developers are blundering nitwits, or this construct is correct.
```
def bind_floating_ip(floating_ip, device):
    """Bind ip to public interface."""
    _execute('ip', 'addr', 'add', str(floating_ip) + '/32',
             'dev', device,
             run_as_root=True, check_exit_code=[0, 2, 254])
```

Third, it does not match with my understanding of how floating ips work in OpenStack. The floating IP definition results in an iptables rule:

```-A nova-network-OUTPUT -d 172.16.255.254/32 -j DNAT --to-destination 192.168.32.2```

If there was a bigger netmask defined in this rule, then all packets in that network would be transferred to that instance, whereas we want packets to other floating IP addresses in the same rangte to go to a different VM instances.

Fourth, the behavior that you describe (every packet from the login node needs to go via 10.141.255.254) we cannot reconstruct. To test, I configured a login VM on a separate node, properly configured it with an IP address in the same range as the virtual cluster. I can see no hops that pass through 10.141.255.254.
```
[root@login ~]# traceroute c001.vc-a
traceroute to c001.vc-a (172.16.0.1), 30 hops max, 60 byte packets
 1  192.168.32.1 (192.168.32.1)  0.270 ms  0.181 ms  0.120 ms
 2  c001.vc-a (172.16.0.1)  0.493 ms  0.420 ms  0.348 ms
```

I think the behavior that you describe (every packet hopping to the controller node) is caused by a misconfiguration somewhere else in the system. My intuition tells me to check if the login node has an ip address in the 172.16.0.0/16 range defined. If it does not it will send each packet to the default gateway first (which is the controller). This would explain the behavior that you see.

**Comment by brunopereira81** [comment]
Heya Hans,

> I can see no hops that pass through 10.141.255.254.

```
[root@controller(150280*) ~]# ip a | grep 192
    inet 192.168.32.1/24 brd 192.168.32.255 scope global br100
[root@controller(150280*) ~]# ssh login-a
Last login: Wed Mar  9 13:51:10 2016 from login-a.vc-a
[root@login ~]# route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         192.168.32.1    0.0.0.0         UG    0      0        0 eth0
192.168.32.0    0.0.0.0         255.255.255.0   U     0      0        0 eth0
[root@login ~]# traceroute c202
traceroute to c202 (172.16.0.202), 30 hops max, 60 byte packets
 1  192.168.32.1 (192.168.32.1)  0.108 ms  0.065 ms  0.044 ms
 2  c202.vc-a (172.16.0.202)  0.228 ms  0.200 ms  0.168 ms
```

So in effect 192.168.32.1 is already your controller.

The NAT rules making sure that the controller is the central point of all your traffic are:

```
DNAT       all  --  *      *       0.0.0.0/0            172.16.255.254       to:192.168.32.3
and back
SNAT       all  --  *      br100   192.168.32.3         0.0.0.0/0            to:172.16.255.254
```

This is from a container point of view:

```
[root@c202 ~]# netstat -r --numeric-hosts
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         172.16.255.254  0.0.0.0         UG        0 0          0 eth1
172.16.0.0      0.0.0.0         255.255.0.0     U         0 0          0 eth1
[root@c202 ~]# logout
Connection to c202 closed.
[root@controller(150280*) ~]# ip a | grep 172.16
    inet 172.16.255.254/16 scope global br100
```

I don't think that you can get away of this with the current way this are setup in projects, but you cannot say "*I think the behavior that you describe (every packet hopping to the controller node) is caused by a misconfiguration somewhere else in the system.*"

No one is trying to call anyone from Openstack *nitwits*, and we sure understand the reason its set the way it is **by default**, but that is made so that Openstack can work with Openstack services, surely you can understand that a /32 default might not be enough for more complex clusters that need to have other addresses the same range with a different subnet.

**Comment by dchirikov** [comment]
Hi Hans,

You are mixing "floating ips bulk create" and "floating ip create"

**Comment by hansthen** [comment]
@dmitry: 

The link you sent only applies to routing. Since the floating IP is always defined on the nova compute node that hosts the VM responding to the virtual machine, routing does not apply to this IP address.  There should be a static route defined on your controller node that will make sure all data in 172.16.0.0/12 is sent correctly over the internal network (via br100) instead of via the default gateway. Can you verify that this static route is defined on your system? Also, as stated, your physical login node will need an address in the 172.16.0.0/16. Can you verify that this is the case? 

I do not think I am mixing floating-ip-bulk-create and floating-ip-create. I use these commands several times each day. I looked up the source code for each of them. If you read my first argument you will see that I used floating-ip-bulk-create to create a network range with /24 and that the end-result is the same.

**Comment by dchirikov** [comment]
I meant, we need to specify prefix for *bulk* commands. Regarding openstack documentation. In trinity_api and cv_install_first_cluster prefix is omitted. I do not think it is correct.

**Comment by hansthen** [comment]
I agree that a floating ip would be nicer if it were created with a netmask. However, without hacking into the source code of openstack, I see no way of changing this. As stated, specifying a netmask when calling floating-ip-bulk-create has no effect on the netmask of the created floating ip. You can try this for yourself if you do not believe me. 

Instead we have defined a static route for 172.16.0.0/12 via br100. 

Also I do not see how the /32 netmask can cause the routing problem you describe in which every packet is routed via the controller. (We are not speaking of packets generated by a login VM on the controller as these by necessity will have to go though the controller to reach the outside world. (Duh)).

**Comment by dchirikov** [comment]
I can't check, but i trust you and can read source code as well. But I suspect not only network prefix on interface is involved here. Could you, please, check iptables rules, ip-related stuff, like routing, netns, xfrm? Mb. some other non-obvious places. For me /32 prefix is really suspicious, as here is no space for broadcast domain.
